engine/runtime/src/main/java/org/enso/interpreter/node/InlineableRootNode.java

```diff
@@ -0,0 +1,56 @@
+package org.enso.interpreter.node;
+
+import com.oracle.truffle.api.CallTarget;
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.nodes.DirectCallNode;
+import com.oracle.truffle.api.nodes.RootNode;
+
+/**
+ * Special interface that allows various {@link RootNode} subclasses to provide
+ * more effective implementation of {@link DirectCallNode}. Used by for example
+ * by {@code BuiltinRootNode}.
+ */
+public interface InlineableRootNode {
+  /**
+   * Provides access to {@link RootNode}. Usually the object shall inherit from
+   * {link RootNode} as well as implement the {@link InlineableRootNode}
+   * interface. This method thus usually returns {@code this}.
+   *
+   * @return {@code this} types as {link RootNode}
+   */
+  public RootNode getRootNode();
+
+  /**
+   * Name of the {@link RootNode}.
+   *
+   * @return root node name
+   */
+  public String getName();
+
+  /**
+   * Override to provide more effective implementation of {@link DirectCallNode}
+   * suited more for Enso aggressive inlining.
+   *
+   * @return a node to {@link DirectCallNode#call(java.lang.Object...) call} the
+   * associated {@link RootNode} - may return {@code null}
+   */
+  public DirectCallNode createDirectCallNode();
+
+  /**
+   * * Obtain a {@link DirectCallNode} for given {@link CallTarget}.Either
+   * delegates to {@link #createDirectCallNode} or uses regular
+   * {@link DirectCallNode#create(com.oracle.truffle.api.CallTarget)} method.
+   * Use for example by {@code ExecuteCallNode}.
+   *
+   * @param target call target with regular or
+   * {@link InlineableRootNode} {@link RootCallTarget#getRootNode()}
+   * @return instance of {@link DirectCallNode} to use to invoke the
+   * {@link RootNode#execute(com.oracle.truffle.api.frame.VirtualFrame)}.
+   */
+  public static DirectCallNode create(RootCallTarget target) {
+    if (target.getRootNode() instanceof InlineableRootNode inRoot && inRoot.createDirectCallNode() instanceof DirectCallNode node) {
```

Comment:
So this is where the magic happens. Neat!

Reply:
`inRoot.createDirectCallNode() instanceof DirectCallNode` seems unnecessary, because the contract of `createDirectCallNode` is to return `DirectCallNode`, right??

It looks like it's a magical check for check if `inRoot.createDirectCallNode()` returns not-`null` and give it a name (`node`). This is very magical. It makes the code suggest it does other things than it really does.

Can't we express the intent explicitly?
```suggestion
    if (target.getRootNode() instanceof InlineableRootNode inRoot) {
      DirectCallNode node = inRoot.createDirectCallNode();
      if (node != null) {
```

Reply:
That was my original version of the code. Then I decided to use new Java features of `instanceof` and make the `if` a one liner. Yes, the second `instanceof` is basically non-`null` check and gives the result a name.

Reply:
Can we use a more explicit code? The meaning is the same, but the `instanceof` trick is IMO very obscure - it obfuscates the original intent behind this code.

Reply:
Btw.

> ...very obscure...

When generics were introduced into Java in JDK 1.5 some people were saying that `List<String>` is _obscure and obfuscates_ the code. The same with lambdas. Usually people get used to such new language features over the time.

Reply:
I don't think I got my point across.

I'm not saying the `instanceof` feautre in _itself_ is obscure. I love this feature, it's cool, I use it wherever I can when it makes sense.

But here you are abusing a **typecheck** to perform a **null-check**. That _is_ obscure. It is a trick. It hides your intentions - the intention was to check for null but you are doing a whole typecheck along with it. And the code reader will be wondering 'why is he checking the type here if the type is known to be correct from the definition...'. I was. I know it's not very hard and one can figure it out, knowing that `instanceof` _also_ does a null check. I'm however very stubborn about this, not about this particular instance, but about the philosophy this brings. Why complicate code unnecessarily? I think the only benefit here is having 1 line instead of 2. But readability is hurt.

IMO the code should try to show the intentions as clearly as possible. If that is not possible, for example due to a more 'cryptic' approach being more efficient - I'm totally OK with that - but then it should be explained.